File: csharp/ExtractJson.cs
// Extract structured invoice data from a Factur-X PDF as JSON using the
// InvoiceXML REST API. Useful for feeding Factur-X invoices into REST
// pipelines, ERPs, or any system that prefers JSON over XML.
//
// Get your API key at: https://www.invoicexml.com/account/authentication
// Full API reference:  https://www.invoicexml.com/docs/api/extract/json
//
// Required NuGet package:
//   dotnet add package Flurl.Http

using Flurl.Http;

public static class ExtractJson
{
    private const string Endpoint = "https://api.invoicexml.com/v1/extract/json";

    public static async Task<string> RunAsync(
        string pdfPath = "invoice.pdf",
        string outputPath = "invoice.json")
    {
        // Get an InvoiceXML API key at: https://www.invoicexml.com/account/authentication
        // Raw key only, without the "Bearer " prefix (WithOAuthBearerToken adds it).
        var apiKey = Environment.GetEnvironmentVariable("INVOICEXML_API_KEY")
            ?? throw new InvalidOperationException(
                "Set INVOICEXML_API_KEY environment variable. " +
                "Get an InvoiceXML API key at " +
                "https://www.invoicexml.com/account/authentication");

        if (!File.Exists(pdfPath))
            throw new FileNotFoundException($"Input PDF not found: {pdfPath}");

        try
        {
            // POST the PDF and receive a JSON response with the parsed EN 16931
            // invoice document (seller, buyer, lines, tax breakdown, totals)
            // under the "invoice" key. Response schema:
            // https://www.invoicexml.com/docs/api/extract/json
            var json = await Endpoint
                .WithOAuthBearerToken(apiKey)
                .PostMultipartAsync(mp => mp
                    .AddFile("file", pdfPath, "application/pdf"))
                .ReceiveString();

            await File.WriteAllTextAsync(outputPath, json);
            Console.WriteLine($"Factur-X data extracted to {outputPath}");
            Console.WriteLine(json);
            return json;
        }
        catch (FlurlHttpException ex)
        {
            var body = await ex.GetResponseStringAsync();
            Console.Error.WriteLine($"InvoiceXML API error {(int?)ex.StatusCode}: {body}");
            throw;
        }
    }
}

File: ruby/extract_json.rb
# Extract structured invoice data from a Factur-X PDF (or a standalone
# CII / UBL XML file) as normalized JSON using the InvoiceXML API.
# Deterministic XML parsing, no AI involved.
#
# Get API key: https://www.invoicexml.com/account/authentication
# Docs:        https://www.invoicexml.com/docs/api/extract/json
#
# Ruby 3.x, standard library only. Set the INVOICEXML_API_KEY
# environment variable to the raw key, without the "Bearer " prefix.
#
# Usage: ruby extract_json.rb [path/to/invoice.pdf]

require "net/http"
require "json"
require "uri"

BASE_URL = "https://api.invoicexml.com"
API_KEY  = ENV.fetch("INVOICEXML_API_KEY")

def api_http(uri)
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = true
  http.open_timeout = 10
  http.read_timeout = 120
  http
end

def extract_invoice(path)
  uri = URI("#{BASE_URL}/v1/extract/json")
  request = Net::HTTP::Post.new(uri)
  request["Authorization"] = "Bearer #{API_KEY}"

  File.open(path, "rb") do |file|
    request.set_form([["file", file]], "multipart/form-data")
    response = api_http(uri).request(request)

    unless response.is_a?(Net::HTTPSuccess)
      warn "InvoiceXML API error #{response.code}: #{response.body}"
      exit 1
    end

    JSON.parse(response.body)
  end
end

data = extract_invoice(ARGV.fetch(0, "facture-facturx.pdf"))

# The invoice document sits under the "invoice" key of the response.
invoice = data["invoice"]
puts "Invoice #{invoice["invoiceNumber"]} from #{invoice.dig("seller", "name")}"
puts "Grand total: #{invoice.dig("totals", "grandTotalAmount")} #{invoice["currency"]}"
